accounts-db/src/read_only_accounts_cache.rs

```diff
@@ -344,46 +318,42 @@ impl ReadOnlyAccountsCache {
     fn evict(
         target_data_size: usize,
         data_size: &AtomicUsize,
+        evict_sample_size: usize,
         cache: &DashMap<ReadOnlyCacheKey, ReadOnlyAccountCacheEntry>,
-        queue: &Mutex<IndexList<ReadOnlyCacheKey>>,
     ) -> u64 {
+        let mut rng = thread_rng();
+
         let mut num_evicts = 0;
         while data_size.load(Ordering::Relaxed) > target_data_size {
-            let Some(&key) = queue.lock().unwrap().get_first() else {
-                // if there are no more entries, we're done
-                break;
+            let key = {
+                let entries = cache.iter().choose_multiple(&mut rng, evict_sample_size);
+                let entry = entries
+                    .iter()
+                    .min_by_key(|entry| entry.value().last_update_time.load(Ordering::Relaxed))
+                    // PANICS: Impossible to happen for a non-empty batch of entries.
+                    .unwrap();
+                // `key()` returns a RefMulti, which holds a read lock on the
+                // `DashMap`. Obtaining an owned key drops the `RefMulti` and
+                // releases the lock.
+                entry.key().to_owned()
```

Comment:
I wish there was a way to obtain an upgradeable lock on the dash map, so I could simply upgrade the read lock before removing. For now the best I can do is to unlock the read lock and then let `do_remove` acquire the write lock.

I also tried whether moving the search for minimum element out of the critical section improves anything in terms of performance, but it doesn't look like. To be precise, I tried:

```rust
let key = {
    let entries = cache
        .iter()
        // DashMap iterator yields a RefMulti, which holds a read
        // lock on the map. We obtain owned values to free that
        // lock.
        // Doing so means that the lines below, searching for the
        // minimum element in the sample, are not the part of
        // critical section. That might result in a race condition,
        // where our picked key might be updated.
        .map(|entry| {
            (
                entry.key().to_owned(),
                entry.value().last_update_time.load(Ordering::Relaxed),
            )
       })
       .choose_multiple(&mut rng, evict_sample_size);
    let (key, _) = entries
       .iter()
       .min_by_key(|(_, last_update_time)| last_update_time)
        // PANICS: Impossible to happen for a non-empty batch of entries.
        .unwrap();
    key.to_owned()
};
```

No difference in benchmarks.

I also tried this dummy, non-Rusty single element pick in a loop, which performs worse:

```rust
let key = {
    let mut min = u32::MAX;
    let mut to_evict = None;
    for _ in 0..evict_sample_size {
        let (key, last_update_time) = cache
            .iter()
            .choose(&mut rng)
            .map(|entry| {
                (
                    entry.key().to_owned(),
                    entry.value().last_update_time.load(Ordering::Relaxed),
                 )
            })
            // PANICS: Impossible to happen for a non-empty batch of entries.
            .unwrap();
        if last_update_time >= min {
            min = last_update_time;
            to_evict = Some(key);
        }
   }

    // PANICS: Impossible to happen for a non-empty batch of entries.
    to_evict.unwrap()
};
```

Reply:
Actually, I think we could try using [`DashMap::shards_mut`](https://docs.rs/dashmap/6.1.0/dashmap/struct.DashMap.html#method.shards_mut) to get to the `RwLock` and then acquire an upgradeable lock. I'm going to try that.